File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticRequest.java
package com.foundgine.core.semantic;
import com.foundgine.core.abstractions.*; import com.foundgine.core.semantic.query.*; import com.foundgine.core.semantic.security.execution.SecurityExecutionContext; import java.util.*;
/** Protocol-neutral description of caller intent. */
public record SemanticRequest(EntityId root,List<SemanticSelection> selections,SemanticQueryOptions options,SecurityExecutionContext security){public SemanticRequest{Objects.requireNonNull(root);selections=List.copyOf(selections==null?List.of():selections);}}

File: src/csharp/Foundgine.Core/Semantic/SemanticRequest.cs
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Query;
using Foundgine.Core.Semantic.Security.Execution;

namespace Foundgine.Core.Semantic;

/// <summary>
/// Protocol-neutral description of what a caller wants. Adapters such as
/// GraphQL translate into this shape; the engine never sees their ASTs.
/// </summary>
public sealed record SemanticRequest(
    EntityId Root,
    IReadOnlyList<SemanticSelection> Selections,
    SemanticQueryOptions? Options = null,
    SecurityExecutionContext? Security = null);

public sealed record SemanticSelection(
    FieldId? Field,
    RelationshipId? Relationship,
    IReadOnlyList<SemanticSelection> Children);
